experimental/cache_analysis.py

```diff
@@ -0,0 +1,139 @@
+import csv
+import seaborn as sns
+
+""" This functions draws two types of plots for cases where each x value has only one y value.
+
+Here it is used to draw plots for stats (ie mean and median of all runs).
+It can be used for drawing each run individually.
+"""
+def draw_single_run_plot(data, name):
+  figure(num=None, figsize=(10, 8), dpi=80, facecolor='w', edgecolor='k')
+
+  x = list(data.keys())
+  y = list(data.values())
+
+  plt.subplot(211)
+  plt.plot(x,y)
+  plt.subplot(212)
+  plt.scatter(x,y)
+  plt.savefig(name)
+  plt.cla()
+
+
+""" This functions draws boxplot for all data points.
+"""
+def draw_boxplot(data, x, name):
+  fig = plt.figure(figsize=(30, 10))
+  ax = fig.add_subplot(111)
+  boxes = []
+  for l in data:
+    c = 0
+    for i in l:
+      boxes.append([x[c], i])
+      c += 1
+  dd = pd.DataFrame(boxes, columns = ["size (bytes)", "time (cycles)"])
+  sns.set_theme(style="whitegrid")
+
+  ax = sns.boxplot(data = dd, showfliers=False, x = 'size (bytes)', y = 'time (cycles)')
+  ax.grid(False)
+  fig = ax.get_figure()
+  fig.savefig(name)
+
+""" This functions draw all data points in a plot in log scale
+"""
+def draw_all_datapoints(data, sizes, name):
+  fig = plt.figure(figsize=(30, 10))
+  ax = fig.add_subplot(111)
+  for l in data:
+    ax.scatter(sizes,l)
+  # start, end = ax1.get_xlim()
+  # ax1.xaxis.set_ticks(np.arange(0, end))
+  # start, end = ax1.get_ylim()
+  # ax1.yaxis.set_ticks(np.arange(0, end))
+  ax.set_yscale('log')
+  ax.set_xscale('log', base = 2)
+  plt.xlabel("size (bytes)")
+  plt.ylabel("time (cycles)")
+  plt.title(" all data points")
+  plt.grid(False)
+  plt.savefig(name)
+  plt.cla()
+
+
+
+file_name = 'cache_line_size_results.csv'  # name of result file
+file_path = '/home/negara/safeside/build/experimental/'  # path to the result file
+
+# read csv file to extract data
+df = None
+try:
+    from google.colab import files
+    # colab
+    uploaded = files.upload()
+    df = pd.read_csv(io.StringIO(
+        uploaded[file_name].decode('utf-8')), header=None)
+except ImportError:
+    # not colab
+    import os
+    df = pd.read_csv(os.path.join(file_path, file_name), header=None)
+
+
+count_row = df.shape[0]  # gives number of row count
+count_col = df.shape[1]  # gives number of col count
+sizes = df[0].unique()  # sizes used for the analysis
+iterations = int(count_row/len(sizes))  # total experiments' repetitions
+print("In total the experiment was repeated ", iterations, " times")
+print("In total ", len(sizes), " different values were tested")
+
+# extract all runs seperately
+all = []
+pre = 0
+for i in range(1, count_row):
+    if (i+1) % len(sizes) == 0:
+        l = df.iloc[pre: i+1]
+        pre = i+1
+        all.append(l)
+
+# extract each size results
+data_per_size = {}
+all_data = []
+for i in range(0, iterations):
+    l = list(all[i][1])
+    l_size = len(l)
+    all_data.append([x for x in l])
+
+    # extracting timings for each size
+    for j in range(0, len(l)):
```

Comment:
nit: this is another good place to use enumerate.

Instead of:

```py
for j in range(len(l)):
  ...
  t.append(l[j])
  ...
```

one can do:

```py
for j, name_goes_here in enumerate(l):
  ...
  t.append(name_goes-here)
  ...
```

Reply:
removed this part as it was unused